loadgen/loadgen.cc

```diff
@@ -1488,4 +1490,109 @@ void QuerySamplesComplete(QuerySampleResponse* responses,
+      val = it->second;
+      return true;
+    }
+    // lookup key with model wildcard first
+    it = kv.find("*." + scenario + "." + key);
+    if (it != kv.end()) {
+      val = it->second;
+      return true;
+    }
+    return false;
+  };
+
+  // dirt simple config parser
+  std::ifstream fss(path);
+  std::string line;
+  int line_nr = 0;
+  int errors = 0;
+  if (!fss.is_open()) {
+    LogDetail([p = path](AsyncDetail & detail) {
+      detail.Error("can't open ", p);
+    });
+    return -ENOENT;
+  }
+  while (std::getline(fss, line)) {
+    line_nr++;
+    std::istringstream iss(line);
+    std::string s, k;
+    int state = 0;
+    while (iss >> s) {
+      if (s == "#" && state != 2) {
+        // done with this line
+        break;
+      }
+      if (state == 2) {
+        // got key and value
+        char *end;
+        kv[k] = std::strtol(s.c_str(), &end, 10);
+      }
+      if (state == 1 && s != "=") {
+        errors++;
+        LogDetail([l = line_nr](AsyncDetail & detail) {
+          detail.Error("error in line ", l);
+        });
+        break;
+      }
+      if (state == 0) k = s;
+      state++;
+    }
+  }
+  if (errors != 0) return -EINVAL;
+
+  // set final values in TestSettings
+  size_t val;
+
+  // keys that apply to all
+  if (lookupkv(model, scenario, "min_duration", val)) min_duration_ms = val;
+  if (lookupkv(model, scenario, "max_duration", val)) max_duration_ms = val;
+  if (lookupkv(model, scenario, "min_query_count", val)) min_query_count = val;
+  if (lookupkv(model, scenario, "max_query_count", val)) max_query_count = val;
+
+  // keys that apply to SingleStream
+  if (lookupkv(model, "SingleStream", "target_latency_percentile", val))
+    single_stream_target_latency_percentile = double(val) / 100.0;
+  if (lookupkv(model, "SingleStream", "target_latency", val))
+    single_stream_expected_latency_ns = val * 1000 * 1000;
+
+  // keys that apply to MultiStream
+  if (lookupkv(model, "MultiStream", "target_latency_percentile", val))
+    multi_stream_target_latency_percentile = double(val) / 100.0;
+  if (lookupkv(model, "MultiStream", "target_qps", val))
+    multi_stream_target_qps = double(val);
+  if (lookupkv(model, "MultiStream", "samples_per_query", val))
+    multi_stream_samples_per_query = int(val);
+  if (lookupkv(model, "MultiStream", "max_async_queries", val))
+    multi_stream_max_async_queries = int(val);
+
+  // keys that apply to Server
+  if (lookupkv(model, "Server", "target_latency_percentile", val))
+    server_target_latency_percentile = double(val) / 100.0;
+  if (lookupkv(model, "Server", "target_latency", val))
+    server_target_latency_ns = val * 1000 * 1000;
+  if (lookupkv(model, "Server", "coalesce_queries", val))
+    server_coalesce_queries = (val == 0) ? false : true;
+  if (lookupkv(model, "Server", "target_qps", val))
+    server_target_qps = double(val);
+
+  // keys that apply to Offline
+  if (lookupkv(model, "Offline", "target_qps", val))
```

Comment:
I suspect some submitters might want to request fractional QPS values.

I wonder if kv should map to a string. Then template or overload lookupkv to do the conversion to integer or double as needed.

Reply:
ok, can do.